rosbag2_transport/src/rosbag2_transport/recorder.cpp

```diff
@@ -395,18 +396,29 @@ const rosbag2_cpp::Writer & RecorderImpl::get_writer_handle()
 
 void RecorderImpl::pause()
 {
-  paused_.store(true);
-  RCLCPP_INFO_STREAM(node->get_logger(), "Pausing recording.");
+  std::lock_guard<std::mutex> state_lock(state_transition_mutex_);
```

Comment:
deadlock here? if this is called from `RecorderImpl::stop` or `RecorderImpl::toggle`?

Reply:
@fujitatomoya Sorry about the deadlocks It was premature and I was a bit in a rush.  
From the last round of review, I was under the impression that we don't really need a mutex lock in such operations as `pause`, `resume` and `toggle_paused` since there is only one atomic variable change inside.  
It seems I have forgotten to rework it.

Reply:
@MichaelOrlov yeah, that is also fine to me if we have agreement on that.

something that i was concerned about that is, https://github.com/ros2/rosbag2/pull/1547#discussion_r1525624065. as you think, probably this is not a problem, so maybe we can remove mutex from `pause` and `resume`?

Reply:
@fujitatomoya I think that ultimately currently we shall protect with mutex only transition between start and stop operation since these operations are not atomic and could cause undefined behavior. All other operations as pause, resume and toggle_paused are atomic operations that change internally only one atomic variable.

I've fix deadlocks in pause(), resume() and toggle_paused() methods see my last commit https://github.com/ros2/rosbag2/pull/1547/commits/d7880740458fa0d96ffb9f8b0185130fdae286e0
- Removed mutex lock in pause(), resume() and toggle_paused() methods
and use only atomic value change for `paused_` variable.
- Renamed `state_transition_mutex_` to the`start_stop_transition_mutex_` because it really intends to protect only those two operations
- Add unit test for `toggle_paused()`.